Editor: Add support for `amp-story-audio-sticker`

AnuragVasanwala opened the pull request:
## Summary

<!-- A brief description of what this PR does. -->
This PR adds support for [amp-story-audio-sticker](https://amp.dev/documentation/components/amp-story-audio-sticker).

## User-facing changes

<!--
Please describe your changes.
Include before/after screenshots or a short video.
-->

https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/d422dc47-b1b7-4d89-b72e-9fa65326f532

https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/86945bcb-6f0b-4bf1-8d94-f18244dca65b

## Testing Instructions

<!--
How can the changes in this PR be verified?
Please provide step-by-step instructions how to reproduce the issue, if applicable.
Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->

<!-- ignore-task-list-start -->
<!-- ignore-task-list-end -->

This PR can be tested by following these steps:

1. Press the Insert Audio Sticker quick action button. Please note that this button is only visible if the story has video with audio/background audio.
2. The `headphone-cat` variant of the sticker is added by default. To change this and other options, select the sticker element and view the style panel.


## Reviews

### Does this PR have a security-related impact?
Yes.
<!-- Examples: new APIs, changes to KSES, etc.  -->

### Does this PR change what data or activity we track or use?
No.
<!-- Examples: changes to telemetry, new third-party APIs -->

### Does this PR have a legal-related impact?
No.
<!-- Examples: new images with unknown sources, new production dependencies with incompatible licenses -->

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.

NOTE: One reference per line!

Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #13440

AnuragVasanwala committed:
✨ Add support for `amp-story-audio-sticker`

AnuragVasanwala committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

swissspidy commented on packages/element-library/src/audio-sticker/layer.js:
```diff
@@ -0,0 +1,27 @@
+function getAudioStickerLayerText(element) {
+  const { alt } = element?.resource || {};
+
+  return alt || __('Audio Sticker', 'web-stories');
```
Audio stickers don't have a `resource` object. Let's remove this.

```suggestion
  return __('Audio Sticker', 'web-stories');
```

AnuragVasanwala replied:
Removed in the upcoming commit.

swissspidy commented on packages/element-library/src/audio-sticker/output.js:
```diff
@@ -0,0 +1,21 @@
+ * limitations under the License.
+ */
+function AudioStickerOutput() {
+  // TODO (@AnuragVasanwala): Need to move this logic here from 'packages/output/src/page.js'.
```
Let me know if you need any help with that.

AnuragVasanwala replied:
Yes, it would be good if you can provide some help here.

Adding anything other than `amp-story-audio-sticker` reflects into story but not Audio Sticker.

swissspidy replied:
What do you mean by "reflects into story"?

AnuragVasanwala replied:
Say, following line renders label to the story:
```js
function AudioStickerOutput() {
  // TODO (@AnuragVasanwala): Need to move this logic here from 'packages/output/src/page.js'.
  return <div>{'AudioStickerOutput'}</div>;
}
```

while this doesn't renders Audio Sticker:
```js
function AudioStickerOutput() {
  // TODO (@AnuragVasanwala): Need to move this logic here from 'packages/output/src/page.js'.
  //return <div>{'AudioStickerOutput'}</div>;
  return (
    <amp-story-grid-layer template="vertical" class="grid-layer">
      <amp-story-audio-sticker size="large" sticker="tape-player" />
    </amp-story-grid-layer>
  );
}
```
Also, even markup of the `amp-story-grid-layer` and `amp-story-audio-sticker` is removed from the story. It seems like it is being sanitised somehow!

swissspidy replied:
> while this doesn't renders Audio Sticker:

Note that the `*Output()` component must not include `amp-story-grid-layer`. `amp-story-grid-layer` is already added by `packages/output/src/page.js`. Otherwise it breaks AMP validation.

----

So, I just pushed some changes to "fix" the output. Turns out the sanitization happens because of a limitation in the  expected document structure right now, as there must not be any divs between the grid layer and the sticker. That seems like an oversight to me, so I asked about context. See https://github.com/ampproject/amphtml/pull/39184#issuecomment-1880878366
Ideally this gets fixed in AMP soon.

In the meantime, if you want to test this locally, you can add an early `return false;` in `Output_Buffer::is_needed()` during development to test the frontend output.

swissspidy replied:
Submitted a PR on the AMP repo here: https://github.com/ampproject/amphtml/pull/39727#issuecomment-1881522325

swissspidy replied:
This is now resolved and the audio sticker should not be removed anymore

swissspidy commented on packages/element-library/src/constants.ts:
```diff
@@ -13,6 +13,16 @@
+  AUDIO_STICKER_ASPECT_RATIOS,
+  AUDIO_STICKER_DEFAULT_PRESET,
+  AUDIO_STICKER_STYLES,
+} from './audio-sticker/constants';
```
Are these only used in the editor? If so, let's define them in the `story-editor` package instead.

swissspidy commented on packages/story-editor/src/components/library/panes/shapes/shapesPane.js:
```diff
@@ -58,6 +58,10 @@ function ShapesPane(props) {
   useRovingTabIndex({ ref: stickersRef });
+
+  const audioStickersRef = useRef(null);
+  useRovingTabIndex({ ref: audioStickersRef });
```
This doesn't seem to be used?

swissspidy commented on packages/element-library/src/elementTypes.js:
```diff
@@ -42,6 +43,11 @@ const elementTypes = [
     ...productElement,
   },
+  {
+    type: 'audio_sticker',
```
Hmm let's rename it to `audioSticker`. We don't typically use snake_case

```suggestion
    type: 'audioSticker',
```

Note: requires some changes in quick actions as well, and anywhere else the element type is used.

AnuragVasanwala replied:
Apologies! I missed that 😅
Corrected in the upcoming commit and will rectify other, if any.

swissspidy commented on packages/element-library/src/audio-sticker/display.js:
```diff
@@ -0,0 +1,95 @@
+      ref={ref}
+      height={elementHeight}
+      width={elementWidth}
+      alt="audio-sticker"
```
This should use the correct label, e.g. Headphone Cat.

swissspidy commented on packages/element-library/src/audio-sticker/display.js:
```diff
@@ -0,0 +1,95 @@
+function AudioStickerDisplay({ element }) {
+  const {
+    id,
+    border,
```
I don't think audio stickers have borders.

swissspidy commented on packages/element-library/src/audio-sticker/constants.js:
```diff
@@ -0,0 +1,100 @@
+
+export const isMaskable = false;
+
+export const isAspectAlwaysLocked = false;
```
Probably needs to be `true`, we don't want users to change the aspect ratio.

AnuragVasanwala committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

AnuragVasanwala committed:
♻️Remove unnecessary attrs

Co-Authored-By: Pascal Birchler <pascal.birchler@gmail.com>

AnuragVasanwala committed:
♻️Fix naming convention to `camelCase`

Co-Authored-By: Pascal Birchler <pascal.birchler@gmail.com>

swissspidy committed:
Use camelCase, change output

swissspidy committed:
Improve styling

AnuragVasanwala committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

AnuragVasanwala committed:
♻️ Move constants to `story-editor` package

AnuragVasanwala committed:
♻️ Update imports for the constants

AnuragVasanwala committed:
⏪ Remove unused ref from `shapesPane`

AnuragVasanwala committed:
♻️ Remove redundant code from `display.js` and add `alt` labels to Audio Sticker Element

AnuragVasanwala committed:
🐛 Add fix for `audioSticker` tap bug, also fixes the outline not appearing bug

Tested with CSS for the muted story state, but it does not work. Thus, applied height 100% to the whole element.

AnuragVasanwala committed:
♻️ Use `ProductMenu` as right click menu for the `audioSticker`

AnuragVasanwala committed:
♻️ Hide Link and Animation panels for `audioSticker` element

AnuragVasanwala committed:
Merge branch 'feat/13440--amp-story-audio-sticker' of https://github.com/GoogleForCreators/web-stories-wp into feat/13440--amp-story-audio-sticker

AnuragVasanwala committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

AnuragVasanwala committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

swissspidy committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

swissspidy committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

Swanand01 committed:
♻️ Remove unused deps and export style presetsto be used in tests

Swanand01 committed:
Add tests for sticker style panels

Swanand01 committed:
Merge branch 'feat/13440--amp-story-audio-sticker' of https://github.com/GoogleForCreators/web-stories-wp into feat/13440--amp-story-audio-sticker

Swanand01 committed:
Add getter for the insert Audio Sticker quick action button

Swanand01 committed:
Add fixture container for the Audio Sticker Type Panel

Swanand01 committed:
Add fixture container for the Audio Sticker Size Panel

Swanand01 committed:
Add fixture container for the Audio Sticker Style Panel

Swanand01 committed:
Add getByText method

Swanand01 committed:
Add getters for the Audio Sticker style panel fixture containers

Swanand01 committed:
Add karma tests for Audio Sticker

Swanand01 committed:
Add test for Audio Sticker output

swissspidy commented on packages/output/src/utils/styles.js:
```diff
@@ -152,6 +152,14 @@ function CustomStyles() {
+                height: 100%;
+              }
+
+              amp-story-grid-layer .i-amphtml-amp-story-audio-sticker-tap-hint {
```
This causes the AMP validation error because we are using `.i-amphtml-...` selectors in CSS.

We need to find another selector to fix the font family.

AnuragVasanwala replied:
@swissspidy I am still getting this validation error in debug panel when previewing story:

<img width="1348" alt="image" src="https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/6f1156b0-ce63-4673-8c19-1bf2f7da8e26">

```
The parent tag of tag 'amp-story-audio-sticker' is 'div', but it can only be 'amp-story-grid-layer'.
```

I think, https://github.com/ampproject/amp-wp/pull/7741 is yet to release with milestone `2.5.3`. Current release is `2.5.2` which was released back in January, 2024.

> This causes the AMP validation error because we are using .i-amphtml-... selectors in CSS.
> We need to find another selector to fix the font family.

I am finding another way to fix this one.

swissspidy replied:
@AnuragVasanwala The debug panel has nothing to do with the AMP plugin. It just means that the public AMP validator used by the runtime is not yet updated. That should just be a matter of time I think.

Updating the AMP plugin is only relevant for our sanitization logic and PHP unit tests.

The AMP plugin release also does not matter as we directly use the develop branch: 

https://github.com/GoogleForCreators/web-stories-wp/blob/e146a96203412f16f98863b7a1de86c095b94522/composer.json#L15

swissspidy commented on packages/story-editor/src/app/quickActions/useQuickActions.js:
```diff
@@ -150,8 +151,21 @@ const useQuickActions = () => {
 
+  const hasAudioAnywhere = useStory(
+    ({ state }) =>
+      state.story.backgroundAudio ||
```
```suggestion
      state.story?.backgroundAudio ||
```

This should mitigate some unrelated test failures

Swanand01 committed:
Update `useQuickActions.js` to mitigate some unrelated test failures

Co-authored-by: Pascal Birchler <pascalb@google.com>

AnuragVasanwala committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

Swanand01 committed:
Change the selector for audio sticker tap hint

Swanand01 committed:
Add null check for state.pages

Swanand01 committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

swissspidy committed:
Patch `@ampproject/toolbox-optimizer`

Swanand01 committed:
Merge branch 'main' of https://github.com/GoogleForCreators/web-stories-wp into feat/13440--amp-story-audio-sticker

Swanand01 committed:
Add `amp-story-audio-sticker` to allowed story components

Swanand01 committed:
Add audio sticker markup to `data_test_filter_kses_allowed_html`

Swanand01 committed:
Add note for the failing output validation test

Swanand01 committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

Swanand01 committed:
Use `waitFor` in the failing test

swissspidy commented on packages/element-library/src/audioSticker/test/output.js:
```diff
@@ -0,0 +1,43 @@
+      box: { width: 1080, height: 1920, x: 50, y: 100, rotationAngle: 0 },
+    };
+
+    // Note: This test will fail until this PR - https://github.com/ampproject/amphtml/pull/39913 is merged.
```
We should use `it.skip` for now and move this comment to the explanation for that. Look at other skipped tests for precedence.

swissspidy commented on packages/story-editor/src/components/canvas/karma/quickActions.karma.js:
```diff
@@ -124,17 +124,20 @@ describe('Quick Actions integration', () => {
-      expect(document.activeElement).toEqual(
-        fixture.editor.canvas.framesLayer.frames[1].node
-      );
+      await waitFor(async () => {
```
You probably only need the `waitFor` around the `expect(fixture.editor.canvas.framesLayer.frames.length).toBe(2);` assertion. `waitFor` is not meant to be wrapped around the whole test.

Please look at other tests for precedence/inspiration.

Swanand01 replied:
The failing assertion is actually - 
```js
expect(document.activeElement).toEqual(
    fixture.editor.canvas.framesLayer.frames[1].node
);
```
I did try wrapping this in `waitFor`, but it still failed. Another way was to use `fixture.events.sleep(100)` after `click()`, but it is essentially the same as using `setTimeout`.

Swanand01 replied:
Going with `fixture.events.sleep(100)` for now

Swanand01 committed:
Skip the output validation test

swissspidy committed:
Fix TypeScript errors

Swanand01 committed:
Update import path for constants

Swanand01 committed:
Replace 'audioSticker' with `ElementType.AudioSticker`

Swanand01 committed:
Use `fixture.events.sleep()` in the failing test

swissspidy commented on packages/story-editor/src/components/panels/design/audioSticker/stickerStyle.js:
```diff
@@ -0,0 +1,121 @@
+              }}
+              aria-label={sprintf(
+                /* translators: %s: Sticker type */
+                __('Sticker Style: %s', 'web-stories'),
```
```suggestion
                __('Sticker Type: %s', 'web-stories'),
```

So we can use the same string twice; less work for translators

Swanand01 replied:
We could, but we'll also have to update many tests for the style panels. These `Sticker Type: %s` and `Sticker Style: %s` labels are used to get and interact with the Sticker Type and Sticker Style buttons, in the style panel. Using the same string twice might fail the tests.

Would you like for me to go ahead and refactor the tests?

Swanand01 replied:
Hi @swissspidy, any updates on this 👆

swissspidy replied:
Apologies. Missed this. Let's leave as-is.

swissspidy commented on packages/story-editor/src/components/panels/design/audioSticker/stickerType.js:
```diff
@@ -0,0 +1,114 @@
+                label
+              )}
+            >
+              <StickerImage src={AUDIO_STICKERS[type]} alt="audio-sticker" />
```
Can't we use `alt={AUDIO_STICKER_LABELS[sticker].label}` here like in other places?

swissspidy commented on packages/story-editor/src/components/panels/design/audioSticker/stickerStyle.js:
```diff
@@ -0,0 +1,121 @@
+            >
+              <StickerImage
+                src={AUDIO_STICKERS[type]}
+                alt="audio-sticker"
```
Can't we use `alt={AUDIO_STICKER_LABELS[sticker].label}` here like in other places?

Swanand01 committed:
Use `AUDIO_STICKER_LABELS` for alt attribute

swissspidy commented on packages/output/src/utils/styles.tsx:
```diff
@@ -152,6 +152,15 @@ function CustomStyles() {
+              amp-story-grid-layer div:has(~ .small) > span,
+              amp-story-grid-layer div:has(~ .large) > span {
+                font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', sans-serif;
+              }
```
This does not appear to be working. Probably because our AMP sanitizer strips this as it doesn't know where it is being used, so it thinks it's unused & removes it.

I think we can solve this by wrapping the `<amp-story-audio-sticker>` in the `AudioStickerOutput` component in an extra `<div className="audio-sticker">`.

Then, here we can do `.audio-sticker { font-family: ... }`

Please test this.

Swanand01 replied:
Yes, tested this and it worked! Pushed commits for the same.

Swanand01 committed:
Wrap output element with a div

Swanand01 committed:
Add `font-family` style using the output div's classname

swissspidy committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

AnuragVasanwala committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

swissspidy commented on packages/element-library/src/audioSticker/test/output.js:
```diff
@@ -0,0 +1,43 @@
+
+describe('AudioSticker output', () => {
+  // eslint-disable-next-line jest/no-disabled-tests -- This test will fail until this PR - https://github.com/ampproject/amphtml/pull/39913 is merged.
+  it.skip('should produce valid AMP output', async () => {
```
@Swanand01 The AMP validator should be updated now. Can you confirm whether this test passes now?

Swanand01 replied:
It passed locally, now waiting for it to pass on this PR too.

Swanand01 replied:
Did not pass:
The extension 'amp-story-audio-sticker' is referenced at version 'undefined' which is an invalid version. Please use a valid version of this extension. (see https://amp.dev/documentation/components/amp-story-audio-sticker)

Swanand01 replied:
@swissspidy How did we fix this the last time? I can't seem to remember 😅

swissspidy replied:
It first I thought it's weird that it passes locally but not on CI. It also did not pass for me locally either.

Then I realized the `@ampproject/toolbox-optimizer` package we use for the AMP validation tests has some heavy disk-based caching. So the information I had locally was super out of date.

86fc09aa5a168a1d4f6f8a4c33228fb4364c5a10 is an attempt at working around this by always using the latest version. Let's see if it works. 🤞

Swanand01 committed:
Re-enable skipped test

swissspidy committed:
Try cache busting by always using latest runtime version

Swanand01 committed:
Merge branch 'main' into feat/13440--amp-story-audio-sticker

swissspidy reviewed (approved):
Let's merge this one for now so we can address any follow-ups!
